File: examples_android/app/src/main/java/com/confirmit/testsdkapp/domain/TriggerCallback.kt
package com.confirmit.testsdkapp.domain

import androidx.appcompat.app.AlertDialog
import com.confirmit.mobilesdk.ConfirmitSDK
import com.confirmit.mobilesdk.trigger.ProgramCallback
import com.confirmit.mobilesdk.trigger.TriggerInfo
import com.confirmit.mobilesdk.ui.SurveyFrameConfig
import com.confirmit.mobilesdk.web.SurveyWebViewFragment
import com.confirmit.testsdkapp.MainApplication
import com.confirmit.testsdkapp.activities.MainActivity
import com.confirmit.testsdkapp.utils.Utils
import com.confirmit.testsdkapp.viewmodels.models.SurveyModel

class TriggerCallback : ProgramCallback {
    override fun onSurveyDownloadCompleted(triggerInfo: TriggerInfo, surveyId: String, exception: Exception?) {
        if (exception != null) {
            val context = ConfirmitSDK.androidContext as MainApplication
            Utils.showError(context.activity!!, "Survey $surveyId download failed", exception.localizedMessage)
        }
    }

    override fun onSurveyStart(config: SurveyFrameConfig) {
        if (OverlayService.isRunning) {
            OverlaySurveyManager.startSurvey(config)
        } else {
            val surveyManager = SurveyManager()
            val context = ConfirmitSDK.androidContext as MainApplication
            surveyManager.launchSurvey(context.activity!!, SurveyModel(config.serverId, config.surveyId), config)
        }
    }

    override fun onScenarioStart(triggerInfo: TriggerInfo, exception: Exception?) {
        exception?.let {
            val context = ConfirmitSDK.androidContext as MainApplication
            Utils.showError(context.activity!!, "Scenario Failed to Start", exception.localizedMessage)
        }
    }

    override fun onScenarioError(triggerInfo: TriggerInfo, exception: Exception) {
        val context = ConfirmitSDK.androidContext as MainApplication
        Utils.showError(context.activity!!, "Scenario Failed", exception.localizedMessage)
    }

    override fun onAppFeedback(triggerInfo: TriggerInfo, data: Map<String, String?>) {
        val context = ConfirmitSDK.androidContext as MainApplication
        context.activity?.let {
            it.runOnUiThread {
                AlertDialog.Builder(it)
                        .setTitle("Write a review!")
                        .setMessage("Would you like to write a review")
                        .setPositiveButton("Write a review!") { _, _ ->
                        }
                        .setNegativeButton("Remind Me Later") { _, _ ->
                        }
                        .show()
            }
        }
    }

    override fun onWebSurveyStart(surveyWebView: SurveyWebViewFragment) {
        val context = ConfirmitSDK.androidContext as MainApplication
        context.activity?.let {
            (it as MainActivity).fetcherListener?.endIdle()
            it.runOnUiThread {
                surveyWebView.setCallback(WebViewCallback())
                surveyWebView.show(it.supportFragmentManager, "SurveyWebViewFragment")
            }
        }
    }
}

File: examples_ios/MyApp/Sources/Domain/TriggerCallback.swift
import ConfirmitMobileSDK
import Foundation
import UIKit

class TriggerCallback: ProgramCallback {
  func onScenarioLoad(triggerInfo: TriggerInfo, error: Error?) {
    // TODO:
  }

  func onSurveyDownloadCompleted(triggerInfo: TriggerInfo, surveyId: String, error: Error?) {
    if let error = error {
      let alert = UIAlertController(title: "Survey \(surveyId) Downloaded Failed", message: error.localizedDescription, preferredStyle: .alert)
      alert.addAction(UIAlertAction(title: "Close", style: .cancel, handler: nil))

      DispatchQueue.main.async {
        if let controller = Utils.getTopController() {
          controller.present(alert, animated: true)
        }
      }
    }
  }

  func onAppFeedback(triggerInfo: TriggerInfo, data: [String: String?]) {
    let alert = UIAlertController(title: "Write a review!", message: "Would you like to write a review for our App?", preferredStyle: .alert)
    alert.addAction(UIAlertAction(title: "Remind Me Later", style: .cancel, handler: nil))
    DispatchQueue.main.async {
      if let controller = Utils.getTopController() {
        controller.present(alert, animated: true)
      }
    }
  }

  func onScenarioStart(triggerInfo: TriggerInfo, error: Error?) {
    if let error = error {
      let alert = UIAlertController(title: "Scenario Failed to Start", message: error.localizedDescription, preferredStyle: .alert)
      alert.addAction(UIAlertAction(title: "Close", style: .cancel, handler: nil))

      DispatchQueue.main.async {
        if let controller = Utils.getTopController() {
          controller.present(alert, animated: true)
        }
      }
    }
  }

  func onScenarioError(triggerInfo: TriggerInfo, error: Error) {
    let alert = UIAlertController(title: "Scenario Failed", message: error.localizedDescription, preferredStyle: .alert)
    alert.addAction(UIAlertAction(title: "Close", style: .cancel, handler: nil))

    DispatchQueue.main.async {
      if let controller = Utils.getTopController() {
        controller.present(alert, animated: true)
      }
    }
  }

  func onSurveyStart(config: SurveyFrameConfig) {
    do {
      let survey = try SurveyModel(serverId: config.serverId, surveyId: config.surveyId)
      let appDelegate = UIApplication.shared.delegate as! AppDelegate
      let navigationController = appDelegate.window!.rootViewController as! UINavigationController
      let controller = navigationController.topViewController
      DispatchQueue.main.async {
        if let controller = controller {
          SurveyManager().launchSurvey(controller: controller, survey: survey, surveyFrameConfig: config)
        }
      }
    } catch {}
  }

  func onWebSurveyStart(surveyWebView: SurveyWebViewController) {
    surveyWebView.getSurveyUrl { url, _ in
      if let url = url {
        // NOTE: Your Survey URL = Use this for custom webview
        print("\(url.url) // \(url.token)")
      }
    }

    if let controller = Utils.getTopController() {
      surveyWebView.delegate = SurveyWebViewDelegate()
      controller.present(surveyWebView, animated: true)
    }
  }
}
